sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -130,36 +153,130 @@ private sealed class EagerCopyingMessageBody : MessageBody
 
             internal EagerCopyingMessageBody(IEnumerable<Data> dataSegments)
             {
+                Append(dataSegments);
+            }
+
+            protected override ReadOnlyMemory<byte> WrittenMemory => _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
+
+            public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+            {
+                return _segments.GetEnumerator();
+            }
+
+            private void Append(IEnumerable<Data> dataSegments)
+            {
+                int length = 0;
+                int numberOfSegments = 0;
+                List<ReadOnlyMemory<byte>> segments = null;
                 foreach (var segment in dataSegments)
                 {
-                    Append(segment);
+                    segments ??= dataSegments is IReadOnlyCollection<Data> readOnlyList
+                        ? new List<ReadOnlyMemory<byte>>(readOnlyList.Count)
+                        : new List<ReadOnlyMemory<byte>>();
+                    ReadOnlyMemory<byte> dataToAppend = segment.Value switch
+                    {
+                        byte[] byteArray => byteArray,
+                        ArraySegment<byte> arraySegment => arraySegment,
+                        _ => ReadOnlyMemory<byte>.Empty
+                    };
+                    length += dataToAppend.Length;
+                    numberOfSegments++;
+                    segments.Add(dataToAppend);
+                }
+
+                if (segments == null)
+                {
+                    return;
+                }
+
+                // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
+                _writer = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
+                _segments = segments;
+
+                for (var i = 0; i < numberOfSegments; i++)
+                {
+                    var dataToAppend = segments[i];
+                    var memory = _writer.GetMemory(dataToAppend.Length);
+                    dataToAppend.CopyTo(memory);
+                    _writer.Advance(dataToAppend.Length);
+                    segments[i] = memory.Slice(0, dataToAppend.Length);
                 }
             }
+        }
 
-            protected override ReadOnlyMemory<byte> WrittenMemory => _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
+        /// <summary>
+        /// TBD because hack
+        /// </summary>
+        private sealed class EagerCopyingMessageBodyWithPooling : MessageBody
+        {
+            private byte[] _buffer;
+            private IList<ReadOnlyMemory<byte>> _segments;
+
+            ~EagerCopyingMessageBodyWithPooling()
```

Comment:
Yeah especially because the finalizer has an impact on the overall application performance as well

Reply:
Honestly I'm still a huge fan of https://github.com/Azure/azure-sdk-for-net/pull/31214. Especially because my original proposal made it opt-in and only for the ServiceBusReceiver callbacks. The argument at the time against the feature was that customers might be keeping messages for longer than the callback and run into issues. I never really agreed with the original argument. My reason was that since the feature is opt-in we could add documentation about what usage scenarios are allowed and what not. And given the array pool documentation explicitly states rented buffers don't necessarily need to be returned, I would much rather resurrect the original PR :) I think it was the cleanest approach and also the most reliable one because the memory is explicitly scoped

Reply:
Plus I also don't understand why such an opt-in feature was considered dangerous. I see it similar as cross entity transactions. Entity transaction is also not a trivial feature to use and has a lot of things to consider from a receiver and sender sharing perspective. Additionally the implicit enlistment of the sender with the receiver even when used outside the receiver flow is also surprising when the feature is not entirely used as it was designed. But such failure scenarios (or shall I say ignorance of reading the feature documentation while enabling) can easily discovered in QA environments or code reviews.

Reply:
So really this attempt here boils down to me trying to convince you all my original approach was a good feature and any other attempt like this one to try to design a bullet proof solution for a very unlikely abuse scenario is too complex to maintain and comes with its own sets of dangers (like having to rely on finalizers)

Reply:
@JoshLove-msft I would love to hear your thoughts at some point for this whole conversation 😊

Reply:
To be clear, I don't think we're fully against that PR - but I'm not sure that the benefits outweigh the cost.

> I never truly agreed with the original argument.

I think the biggest concern was ROI for the additional complexity.  It adds not only complexity for customers (though opt-in) but also in the implementation.  IIRC, our biggest bottleneck for that flow still comes down to the AMQP library serialization/deserialization with its object model being heavy for how we use it.  

> customers might be keeping messages for longer than the callback and due to that run into issues.  we could add documentation about what usage scenarios are allowed

They will, yes.   Inevitably, they will.  The support cost for that additional complexity also concerns me.  I've seen way too many cases in which documented guidance was ignored and which takes a non-trivial amount of investigation to track down due to lack of information or misconceptions.  This is another complexity cost that I think we'd need to be completely sure is worth it before moving forward.

> array pool documentation explicitly states rented buffers don't necessarily need to be returned

To be fair, that is true.  However, the experience on constrained environments - such as many containerized applications - concerns me.

Reply:
I like this discussion! By the way all the arguments I bring up are meant to have a friendly argument about this topic and by no means meant to be disrespectful. I have a lot of empathy for the maintenance cost, complexity, support burden etc because I have been in similar situations.

When we look at the original PR the complexity increase of the code is very minimal. I'd argue the biggest complexity is already today hidden in the various body management strategies. Supporting buffering as opt in would only add a marginal extension to that.

I think the argument of the Amqp lib being a bottleneck is factual and true but it is also an argument that can always be pulled. When it comes to saving allocations for the allocations this lib is responsible for I think it is worth having the discussion regardless of other allocations caused by "third parties“. Copying the memory into the received message is a safety technique that totally makes sense for the general purpose case for both the processor and the receiver. For the processor though because there is an explicit scope of the event handler liberating the safety constraint for performance reasons as an opt-in is a good thing because the allocation costs can be set to a fixed stable costs at runtime without impacting GC overhead anymore.

FYI Rabbitmq client uses this model as a default even and it is the responsibility of the consumer to copy when necessary. I do think though that is a very advanced way of pushing the responsibility to the user code that I feel is definitely out of scope for this library.

So for me the burden for supportability would be so far the "best" argument to not allow optional body buffering for processors.

Again thanks for taking the time reading my ramblings

Reply:
> I like this discussion! By the way all the arguments I bring up are meant to have a friendly argument about this topic and by no means meant to be disrespectful. I have a lot of empathy for the maintenance cost, complexity, support burden etc because I have been in similar situations.

We're all friends here.   No ill intent was assumed, nor will it ever be.  😄

Reply:
> I think the argument of the Amqp lib being a bottleneck is factual and true but it is also an argument that can always be pulled. 

True enough.  

> Honestly I'm still a huge fan of #31214.  Especially because my original proposal made it opt-in and only for the ServiceBusReceiver callbacks. The argument at the time against the feature was that customers might be keeping messages for longer than the callback and due to that run into issues.

I think you're referring to #20320 here, no?    The "I'm holding this message longer than the scope of the callback" is legit and something that I've seen occasionally in application code.  The opt-in makes it caveat emptor, though.   So long as we have that, and it defaults to "nope" - I'm good with it.   I prefer that over the reliance on finalization. 

> FYI Rabbitmq client uses this model as a default even and it is the responsibility of the consumer to copy when necessary. I do think though that is a very advanced way of pushing the responsibility to the user code that I feel is definitely out of scope for this library.

Definitely agree that's not something we'd want to do by default.   If we accept the "if you turn the feature on, you're responsible for coping the body if you use it beyond the callback", then its good to know it's not unprecedented.

> I'd argue the biggest complexity is already today hidden in the various body management strategies.  So for me the burden for supportability would be so far the "best" argument to not allow optional body buffering for processors.

I agree with both of those.  _(hell, the former bit me hard last week)_    

I'm on the fence here, truthfully.  We are near the end of the low-hanging fruit for improving things.   I'm not sure if our next step should be working towards these kinds of improvements or starting to more directly contribute to the AMQP lib itself.

@JoshLove-msft: I'd appreciate your thoughts.

Reply:
Yeah sorry I fat fingered things on my phone. https://github.com/Azure/azure-sdk-for-net/pull/20320 it was indeed

Reply:
@JoshLove-msft any thoughts on the above?

Reply:
We brought up this approach several months back with the .NET architects and they felt that a paradigm shift to allow reusing buffers wasn't warranted. Basically, they felt that there were plenty of optimizations that could be made in the AMQP lib before we would even consider doing something that would impact the public surface. But I don't want to send you down the rabbit hole of trying to fix everything in the AMQP lib 😆 because my understanding is that even if we did that, this feature would still not be likely to be approved as it breaks with the patterns we use in all other Azure client libraries.